### ISSUE TITLE
RuleForEach with SetValidator using a custom validator with When clause, execute condition only for the first item

### ISSUE THREAD

**Issue by sdargaud** [FluentValidation/FluentValidation#970]
### System Details

- FluentValidation version: 8.1.0
- (Web) Framework version: ASP.NET Core 2.2 / .NET Core 2.2 Console application

### Issue Description

When using .RuleForEach().SetValidator with a custom validator containing a When operator, the condition is run only for the first item, and the When rules are applied to all subsequent items if true for the first one.

### Expected behavior

Condition is run for each item in the collection so that not all the rules are run for each item.

### Repro steps

```
public class ForEachWithWhenTests
{
    public class Person
    {
        public IList<Address> Addresses { get; set; }
    }

    public class Address
    {
        public int Id { get; set; }
    }

    public class AddressValidator : AbstractValidator<Address>
    {
    }

    public class PersonValidator : AbstractValidator<Person>
    {
    }

    [Fact]
    public void When_condition_executed_for_each_instance_of_RuleForEach()
    {
        var person = new Person
        {
            Addresses = new List<Address>
            {
                new Address{ Id = 1},
                new Address{ Id = 0}
            }
        };

        var addressValidator = new AddressValidator();
        int executions = 0;
        addressValidator.When(a =>
        {
            executions++;
            return a.Id != 0;
        }, () => { addressValidator.RuleFor(a => a.Id).Equal(1); });
        var personValidator = new PersonValidator();
        personValidator.RuleForEach(p => p.Addresses).SetValidator(addressValidator);
        var validationResult = personValidator.Validate(person);
        Assert.True(validationResult.IsValid);
        Assert.Equal(2, executions);
    }
}
```

By the way, it works in 8.0.101.

Thanks for your help !

**Comment by JeremySkinner** [comment]
This is related to #934 which was an optimisation done so root-level `When` conditions are only run a single time, but this does indeed break `RuleForEach`. I'll aim to have a fix later today. 

For now you can work around this by using the inline version of `When`.

**Comment by JeremySkinner** [comment]
I think the fix should be just to include the object being validated's hash code in the cache key used by the condition.

**Comment by JeremySkinner** [comment]
Fix is now deployed to nuget as 8.1.1. Could you give it a try and confirm it’s working ok?
